zcash_client_backend/src/fees/common.rs

```diff
@@ -385,106 +381,99 @@ where
+                    (total_in - total_out_with_max_fee).unwrap_or(Zatoshis::ZERO),
                 )
-                .unwrap(),
-            );
+            }));
 
             // If we don't have as many change outputs as we expected, recompute the fee.
-            let (fee_with_change, excess_fee) =
-                if usize::from(split_count) < target_change_counts.total_shielded() {
-                    let new_fee_with_change = cfg
-                        .fee_rule
-                        .fee_required(
-                            cfg.params,
-                            target_height,
-                            transparent_input_sizes,
-                            transparent_output_sizes,
-                            sapling_input_count,
-                            sapling_output_count(if change_pool == ShieldedProtocol::Sapling {
-                                usize::from(split_count)
-                            } else {
-                                0
-                            })?,
-                            orchard_action_count(if change_pool == ShieldedProtocol::Orchard {
-                                usize::from(split_count)
-                            } else {
-                                0
-                            })?,
-                        )
-                        .map_err(|fee_error| ChangeError::StrategyError(E::from(fee_error)))?;
-                    (
-                        new_fee_with_change,
-                        (fee_with_change - new_fee_with_change).unwrap_or(Zatoshis::ZERO),
+            let total_fee = if split_count < target_change_count {
+                cfg.fee_rule
+                    .fee_required(
+                        cfg.params,
+                        target_height,
+                        transparent_input_sizes,
+                        transparent_output_sizes,
+                        sapling_input_count,
+                        sapling_output_count(if change_pool == ShieldedProtocol::Sapling {
+                            split_count
+                        } else {
+                            0
+                        })?,
+                        orchard_action_count(if change_pool == ShieldedProtocol::Orchard {
+                            split_count
+                        } else {
+                            0
+                        })?,
                     )
-                } else {
-                    (fee_with_change, Zatoshis::ZERO)
-                };
+                    .map_err(|fee_error| ChangeError::StrategyError(E::from(fee_error)))?
+            } else {
+                max_fee
+            };
+
+            let total_out = (subtotal_out + total_fee).ok_or_else(overflow)?;
+            let total_change =
+                (total_in - total_out).ok_or_else(|| ChangeError::InsufficientFunds {
+                    available: total_in,
+                    required: total_out,
+                })?;
```

Comment:
Please write a test that would fail if this check was not done, i.e. when there are insufficient funds to add a single change note. (If that case already failed with existing tests, that's sufficient.)

Reply:
Done, added to the suggestions commit.